src/celeritas/neutron/interactor/detail/InvariantQ2Sampler.hh

```diff
@@ -0,0 +1,363 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/neutron/interactor/detail/InvariantQ2Sampler.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <cmath>
+
+#include "corecel/Macros.hh"
+#include "corecel/Types.hh"
+#include "corecel/math/Algorithms.hh"
+#include "celeritas/Quantities.hh"
+#include "celeritas/Types.hh"
+#include "celeritas/mat/IsotopeView.hh"
+#include "celeritas/random/distribution/BernoulliDistribution.hh"
+#include "celeritas/random/distribution/UniformRealDistribution.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Sample the momentum transfer (\f$ Q^{2} = -t \f$) of the neutron-nucleus
+ * elastic scattering.
+ *
+ * \note This performs the same sampling routine as in Geant4's
+ *  G4ChipsElasticModel and G4ChipsNeutronElasticXS where the differential
+ *  cross section for quark-exchange and the amplitude of the scattering are
+ *  parameterized in terms of the neutron momentum (GeV/c) in the lab frame.
+ */
+class InvariantQ2Sampler
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using Mass = units::MevMass;
+    //!@}
+
+  public:
+    // Construct with shared and target data, and the neutron momentum
+    inline CELER_FUNCTION InvariantQ2Sampler(NeutronElasticRef const& shared,
+                                             IsotopeView const& target,
+                                             real_type neutron_p);
+
+    // Sample the momentum transfer
+    template<class Engine>
+    inline CELER_FUNCTION real_type operator()(Engine& rng);
+
+  private:
+    //// TYPES ////
+
+    using UniformRealDist = UniformRealDistribution<real_type>;
+
+    //// DATA ////
+
+    ChipsDiffXsCoefficients::ChipsArray const& par_;
+
+    // Mass values in GeV
+    real_type neutron_mass_;
+    real_type target_mass_;
+
+    // Atomic mass number (A) of the target
+    real_type A_;
+    bool heavy_target_{false};
+    // Momentum magnitude of the incident neutron (GeV/c)
+    real_type neutron_p_;
+    // Maximum momentum transfer for the elastic scattering
+    real_type max_Q2_;
+    // Parameters for the CHIPS differential cross section
+    ExchangeParameters par_Q2_;
```

Comment:
Maybe rename `x_Q2` to `x_q_sq_`. I think we don't use capital letters for variables by standard, right?

Reply:
yeah exactly. No capital letters or math symbols; use english or shortened english words (`mom_sq_`).

Reply:
Again, Q^2 has a special (?) meaning for the momentum transfer square - so, `mom_sq_` is somewhat misleading too.  As it is equivalent to the Mandalstam variable `t`, I was kind of using `max_t_` or `par_t_`, but was curious that how many will catch the real meaning of `t`.  To me, I would like it keep it as x_q2_ at least for now unless you have a strong objection.

Reply:
I agree that $Q^2$ has a special meaning. Even more so since this is a four-momentum and not the three-momentum $\vec{p}$, so I too prefer to keep it as Q. But since we lower case everything, and any private/public variable is well documented, I think `q_sq` looks pretty good (and maybe cleaner to read than something `four_mom_transf_sq`) and maintains consistency throughout the codebase.

I also think that moving from $Q2$ to `q_sq` (or `QSq` for classes) is a smaller compromise than other examples in the past. We never use things like `m_eplus` or `E_eplus`. Even constants are lengthened for clarity: `hbar`, which would be universally agreed upon---we call it `hbar_planck`.

Reply:
For the Mandelstam `t`, since it's also a four-momentum notation, how about something like `mandelstam_t` or `lorentz_inv_t`, since it is a lorentz invariant?

Reply:
My concern with capitalization is that in contexts like deep inelastic scattering, they're often defined as being opposite signs, e.g. if $p\_\mu$ is the incident 4-momenta of an electron and $p'_\mu$ is its final 4-momenta, then $q\_\mu = p\_\mu - p'\_\mu$ is the boson exchange momentum and $Q^2 = -q^2 \geq 0$ in the mostly minus metric. If we decide to use `q` or `Q` then we should definitely be explicit with denoting its sign. Mandelstam variables would also be a decent alternative.

I also strongly favor using `_sq` postfix instead of `2`. I found Geant's arrays of `A23` and `A17` confusing before learning they were just cached values of $A^{2/3}$ and $A^{1/7}$ resp.

Reply:
Okay, it is good that @hhollenb is taking QFT and bring us some new conventions!  Of course, we can rename anything as needed later - I prefer not to refactor codes too early until all physics verification with respect to Geant4 are done even though the right choice of naming convention is still a good practice and should be imposed no matter what from the beginning.